tests/Unit/FactoryTest.php

```diff
@@ -638,26 +632,61 @@ public function testCreateNonExistsClass(): void
         $factory->create('NonExistsClass');
     }
 
-    public function testContainerInterfaceWithFactory(): void
+    public function testCreateContainerInterfaceWithoutContainer(): void
     {
-        $factory = new Factory(null, ['x' => new stdClass()]);
+        $factory = new Factory();
 
-        $container = $factory->create(ContainerInterface::class);
+        $this->expectException(NotFoundException::class);
+        $factory->create(ContainerInterface::class);
+    }
 
-        $this->assertTrue($container->has('x'));
-        $this->assertFalse($container->has('y'));
+    public function testCreateContainerInterfaceWithContainerWithoutContainerInterface(): void
+    {
+        $factory = new Factory(new SimpleContainer());
+
+        $this->expectException(NotFoundException::class);
+        $factory->create(ContainerInterface::class);
     }
 
-    public function testContainerInterfaceWithContainer(): void
+    public function testCreateContainerInterfaceWithDefinitionWithoutContainer(): void
     {
-        $factory = new Factory(
-            new SimpleContainer(['x' => new stdClass()])
+        $factory = new Factory(null, [
+            ContainerInterface::class => new SimpleContainer(),
+        ]);
+
+        $containerFromFactory = $factory->create(ContainerInterface::class);
+
+        $this->assertInstanceOf(SimpleContainer::class, $containerFromFactory);
+    }
+
+    public function testCreateContainerInterfaceWithContainerWithContainerInterface(): void
+    {
+        $container = new SimpleContainer(
+            [],
+            static function (string $id) use (&$container) {
+                if ($id === ContainerInterface::class) {
+                    return $container;
+                }
+                throw new \Yiisoft\Test\Support\Container\Exception\NotFoundException($id);
+            }
         );
 
-        $container = $factory->create(ContainerInterface::class);
+        $factory = new Factory($container);
+
+        $this->expectException(NotFoundException::class);
+        $factory->create(ContainerInterface::class);
+    }
+
+    public function testCreateContainerInterfaceWithDefinitionAndContainer(): void
```

Comment:
```suggestion
    public function testCreateDependencyWithDefinitionAndContainer(): void
```

We'd better use non-container for it.

Reply:
Done